pkg/xray/watches.go

```diff
@@ -0,0 +1,267 @@
+	Active      bool   `json:"active"`
+}
+
+type WatchFilterValue struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+type WatchFilter struct {
+	Type  string `json:"type"`
+	Value string `json:"value"`
+}
+
+type WatchProjectResource struct {
+	Type            string        `json:"type"`
+	BinaryManagerId string        `json:"bin_mgr_id"`
+	Filters         []WatchFilter `json:"filters"`
+	// Watch a repo
+	Name string `json:"name"`
+}
+
+type WatchProjectResources struct {
+	Resources []WatchProjectResource `json:"resources"`
+}
+
+type WatchAssignedPolicy struct {
+	Name string `json:"name"`
+	Type string `json:"type"`
+}
+
+type Watch struct {
+	GeneralData      WatchGeneralData      `json:"general_data"`
+	ProjectResources WatchProjectResources `json:"project_resources"`
+	AssignedPolicies []WatchAssignedPolicy `json:"assigned_policies"`
+	WatchRecipients  []string              `json:"watch_recipients"`
+}
+
+func unpackWatch(d *schema.ResourceData) Watch {
+	watch := Watch{}
+
+	gd := WatchGeneralData{
+		Name: d.Get("name").(string),
+	}
+	if v, ok := d.GetOk("description"); ok {
+		gd.Description = v.(string)
+	}
+	if v, ok := d.GetOk("active"); ok {
+		gd.Active = v.(bool)
+	}
+	watch.GeneralData = gd
+
+	pr := WatchProjectResources{}
+	if v, ok := d.GetOk("watch_resource"); ok {
+		r := []WatchProjectResource{}
+		for _, res := range v.(*schema.Set).List() {
+			r = append(r, unpackProjectResource(res))
+		}
+		pr.Resources = r
+	}
+	watch.ProjectResources = pr
+
+	ap := []WatchAssignedPolicy{}
+	if v, ok := d.GetOk("assigned_policy"); ok {
+		policies := v.(*schema.Set).List()
+		for _, pol := range policies {
+			ap = append(ap, unpackAssignedPolicy(pol))
+		}
+	}
+	watch.AssignedPolicies = ap
+
+	var watchRecipients []string
+
+	if v, ok := d.GetOk("watch_recipients"); ok {
+		recipients := v.(*schema.Set).List()
+		for _, watchRec := range recipients {
+			watchRecipients = append(watchRecipients, watchRec.(string))
+		}
+	}
+	watch.WatchRecipients = watchRecipients
+
+	return watch
+}
+
+func unpackProjectResource(rawCfg interface{}) WatchProjectResource {
+	resource := new(WatchProjectResource)
+
+	cfg := rawCfg.(map[string]interface{})
+	resource.Type = cfg["type"].(string)
+
+	if v, ok := cfg["bin_mgr_id"]; ok {
+		resource.BinaryManagerId = v.(string)
+	}
+	if v, ok := cfg["name"]; ok {
+		resource.Name = v.(string)
+	}
+
+	if v, ok := cfg["filter"]; ok {
+		resourceFilters := unpackFilters(v.([]interface{}))
+		resource.Filters = resourceFilters
+	}
+
+	return *resource
+}
+
+func unpackFilters(list []interface{}) []WatchFilter {
+	var filters []WatchFilter
+
+	for _, raw := range list {
+		filter := new(WatchFilter)
+		f := raw.(map[string]interface{})
+		filter.Type = f["type"].(string)
+		filter.Value = f["value"].(string)
+		filters = append(filters, *filter)
+	}
+
+	return filters
+}
+
+func unpackAssignedPolicy(rawCfg interface{}) WatchAssignedPolicy {
+	policy := new(WatchAssignedPolicy)
```

Comment:
I forgot if this pointer is necessary. If not, please convert it.

Reply:
converted, it was unnecessary